senders/pushover/pushover.go

```diff
@@ -71,29 +74,85 @@ func (sender *Sender) makePushoverMessage(events moira.NotificationEvents, conta
+func (sender *Sender) buildMessage(events moira.NotificationEvents, throttled bool, trigger moira.TriggerData) string {
+	var message strings.Builder
+	desc := trigger.Desc
+	htmlDesc := string(blackfriday.Run([]byte(desc)))
+	descLen := len([]rune(desc))
+	htmlDescLen := len([]rune(htmlDesc))
+	charsForHTMLTags := htmlDescLen - descLen
+	var eventsString string
+	for _, event := range events {
+		line := fmt.Sprintf("%s: %s = %s (%s to %s)", event.FormatTimestamp(sender.location), event.Metric, event.GetMetricValue(), event.OldState, event.State)
+		if len(moira.UseString(event.Message)) > 0 {
+			line += fmt.Sprintf(". %s\n", moira.UseString(event.Message))
+		} else {
+			line += "\n"
 		}
-		message.WriteString(fmt.Sprintf("%s: %s = %s (%s to %s)", event.FormatTimestamp(sender.location), event.Metric, event.GetMetricValue(), event.OldState, event.State))
+		eventsString += line
+	}
+	if throttled {
+		eventsString += "\nPlease, fix your system or tune this trigger to generate less events."
+	}
+	eventsStringLen := len([]rune(eventsString))
+	if htmlDescLen+eventsStringLen < msgLimit {
+		// Use both completely
+		message.WriteString(htmlDesc)
+		message.WriteString(eventsString)
+		return message.String()
+	}
+	if htmlDescLen > msgLimit/2 && eventsStringLen > msgLimit/2 {
+		// Trim both desc and events to half the message size each
+		desc = desc[:msgLimit/2-charsForHTMLTags-10] + "...\n"
+		htmlDesc = string(blackfriday.Run([]byte(desc)))
+		eventsString = sender.buildEventsString(events, msgLimit/2, throttled)
+	} else if htmlDescLen > msgLimit/2 {
+		// Trim the desc to the chars left after using the whole events string
+		charsForDesc := msgLimit - eventsStringLen
+		desc = desc[:charsForDesc-charsForHTMLTags-10] + "...\n"
+		htmlDesc = string(blackfriday.Run([]byte(desc)))
+	} else if eventsStringLen > msgLimit/2 {
+		// Trim the events string to the chars left after using the whole desc
+		charsForEvents := msgLimit/2 - htmlDescLen
+		eventsString = sender.buildEventsString(events, charsForEvents, throttled)
+	}
+	message.WriteString(htmlDesc)
+	message.WriteString(eventsString)
+	return message.String()
+}
+func (sender *Sender) buildEventsString(events moira.NotificationEvents, charsForEvents int, throttled bool) string {
+	charsForThrottleMsg := 0
+	if throttled {
+		charsForThrottleMsg = 70
+	}
+	charsLeftForEvents := charsForEvents - charsForThrottleMsg
+	var eventsString string
+	eventsLenLimitReached := false
+	eventsPrinted := 0
+	for _, event := range events {
+		line := fmt.Sprintf("%s: %s = %s (%s to %s)", event.FormatTimestamp(sender.location), event.Metric, event.GetMetricValue(), event.OldState, event.State)
 		if len(moira.UseString(event.Message)) > 0 {
-			message.WriteString(fmt.Sprintf(". %s\n", moira.UseString(event.Message)))
+			line += fmt.Sprintf(". %s\n", moira.UseString(event.Message))
 		} else {
-			message.WriteString("\n")
+			line += "\n"
 		}
+		if len([]rune(eventsString+line)) > charsLeftForEvents {
```

Comment:
I'm returning again to this line. Yes, as you've correctly mentioned in a previous discussion, `len([]rune(somestring))` works fast. But the problem is with `+`. When you do `eventsString+line`, Go allocates another string of length `len(eventsString)+len(line)` and _copies_ contents of two variables there. A whole lot of copying just to count the sum of lengths.

`len([]rune(eventsString)) + len([]rune(line))` should be much faster, since this code does not copy anything.